### ISSUE TITLE
How to compile to a shared lib

### ISSUE THREAD

**Issue by jkroso** [memononen/nanovg#205]
Sorry to ask a support question here but I'm stumped. I can compile it to a shared lib easy enough. All I need to do is s/StaticLib/SharedLib/ in the premake file. But it ends up missing the `nvgCreateGL3` functions. To get them in there I tried adding "NANOVG_GL3_IMPLEMENTATION" to the set of defines in the premake file but it still ends up missing the functions. Any help would be much appreciated.

[![Bountysource](https://www.bountysource.com/badge/issue?issue_id=9251131)](https://www.bountysource.com/issues/9251131-how-to-compile-to-a-shared-lib?utm_source=9251131&utm_medium=shield&utm_campaign=ISSUE_BADGE)

**Comment by memononen** [comment]
The GL backend is a header file library only. If you want to create a static lib you could add a stub C file:
```C
/* gl3backend.c */
#define NANOVG_GL3_IMPLEMENTATION
#include "nanovg_gl.h"
```

The single header file library idea came from Sean: https://github.com/nothings/stb

**Comment by jkroso** [comment]
Ok so to try that I created the "gl3backend.c" file and added an entry in the premake file for it. But when I run `make gl3backend` it ends up failing in the compile step with "unknown type name 'NVGcontext'" and a bunch of other similar errors. `cc -Isrc -o gl3backend.o -c src/gl3backend.c` produces the same output if you want to take a quick look.

**Comment by memononen** [comment]
My bad, you'll need to include `nanovg.h` and your choice of GL header too.

**Comment by jkroso** [comment]
So the stub file should look like this:

```c
#define NANOVG_GL3_IMPLEMENTATION
#include "nanovg.h"
#include "nanovg_gl.h"
#include "nanovg_gl3.h"
```

Where "nanovg_gl3.h" is coming from the "obsolete" folder. Correct?

**Comment by memononen** [comment]
Nope, nothing should come from the obsolete folder. The idea is to make a C file that includes everthing to make the GL backend to compile. With the setup I use for the demos, it probably would like this:

```C
/* gl3backend.c */
#ifdef NANOVG_GLEW
#  include <GL/glew.h>
#endif
#ifdef __APPLE__
#	define GLFW_INCLUDE_GLCOREARB
#endif
#include <GLFW/glfw3.h>
#include "nanovg.h"
#define NANOVG_GL3_IMPLEMENTATION
#include "nanovg_gl.h"
```

**Comment by jkroso** [comment]
Yuus finally got it to compile and link by looking more closely at how the demos worked. Thanks for your help. Would you like me to PR with stuff to make compiling shared libs easy or is it unusual to need them?

**Comment by memononen** [comment]
I think it is unusual to use it that way. I've been thinking of splitting the backend to .c/.h, which should make it easier make a shared lib when needed.

**Comment by memononen** [comment]
Meh... hit wrong button, did not mean to close this :)

**Comment by jkroso** [comment]
Yea and this issue probably documents it well enough anyway. Feel free to claim to bounty too. Otherwise it just sits in limbo. Thanks for your help and for this project

**Comment by thor-the-norseman** [comment]
It's touchy to use these header-only libraries in production code. I think it would be nicer if functions that aren't called in tight loops had their implementations moved to separate .c files.
